java/java.j2seplatform/src/org/netbeans/modules/java/j2seplatform/platformdefinition/J2SEPlatformDefaultJavadocImpl.java

```diff
@@ -116,6 +77,43 @@ public Collection<URI> getDefaultJavadoc(@NonNull final JavaPlatform platform) {
         return Collections.emptyList();
     }
 
+    private static String computeJavaDocURL(String version) {
+        switch (version) {
+            case "1.0": // NOI18N
+            case "1.1": // NOI18N
+            case "1.2": // NOI18N
+            case "1.3": // NOI18N
+            case "1.4": return null; // NOI18N
+            case "1.5": return "https://docs.oracle.com/javase/1.5.0/docs/api/"; // NOI18N
+            case "1.6": return "https://docs.oracle.com/javase/6/docs/api/"; // NOI18N
+            case "1.7": return "https://docs.oracle.com/javase/7/docs/api/"; // NOI18N
+            case "1.8": return "https://docs.oracle.com/javase/8/docs/api/"; // NOI18N
+        }
+        try {
+            int feature = Integer.parseInt(version);
+            if (feature >= 9) {
+                int latestGA = computeLatestGAVersion();
+                if (feature <= latestGA) {
+                    return "https://docs.oracle.com/en/java/javase/" + feature + "/docs/api/"; // NOI18N
+                } else if (feature <= latestGA + 3) {
+                    return "https://download.java.net/java/early_access/jdk" + feature + "/docs/api/"; // NOI18N
+                }
+            }
+        } catch (IllegalArgumentException ignore) {}
+        LOG.log(Level.WARNING, "unrecognized Java spec version: {0}", version); // NOI18N
+        return null;
+    }
+
+    /**
+     * Computes the feature version of the latest generally available JDK release.
+     */
+    private static int computeLatestGAVersion() {
+        // timezone shouldn't matter since the accuracy is worse than a day
+        LocalDate jdk9 = LocalDate.of(2017, Month.SEPTEMBER, 21); // start of 6 month schedule
+        int latest = 9 + (int) (ChronoUnit.MONTHS.between(jdk9, LocalDate.now()) / 6);
+        return Math.max(latest, SourceVersion.latest().ordinal()); // in case system time is wrong, use nb-javac version as lower bound
+    }
```

Comment:
@neilcsmith-net added lower bound

Reply:
@mbien this should be good for releases, but will potentially break in nightly builds where nb-javac is ahead of GA (eg. #6206 )?  Not a blocker IMO though.

Reply:
right. That was the main reason I was hesitant to add the lower bound, Runtime JDK version would have been similar though. Since this would confuse the logic only when NB is tested on EA builds.

unless we add "lowerbound - 1" as safety margin there is no good solution to this.

you think I should remove Math.max() again?

Reply:
No, I think it's fine for releases, which is the most important thing!

The only foolproof ways I can think of are to ping the docs site (which obviously needs to happen lazily) or redirect via netbeans.a.o which we can update externally.

Reply:
the proper way of doing this IMO would be to simply ask the JDK:
```
bin/java -version
openjdk version "22-ea" 2024-03-19
```
or:
`System.out.println(Runtime.version().pre());`

```
--- exec:3.1.0:exec (default-cli) @ lambda_expressions ---
Optional[ea]
```
but until then we can stick to the calendar version since I believe it is working pretty well as workaround

Reply:
Or we just query the passed in `JavaPlatform` version / system properties!  If it's an `ea` platform, link to `ea` javadocs.

Reply:
quickly checkend and both:
```
        System.out.println(platform.getSystemProperties());
        System.out.println(platform.getProperties());
```
are empty when the method is called. But yes that would be also an option, the are also jdk CLI tools which can query properties, but they start a JVM too.

Reply:
Curious!  I just ran it in a debugger, and they're not empty there.  Platform registration already runs the JDK to extract the system properties.

Reply:
i think i found the reason why they are empty. one sec

Reply:
so there is a proxy object for JavaPlatform:
https://github.com/apache/netbeans/blob/a9aeefafc04e2d79a4702be88d890324365989bd/java/java.platform/src/org/netbeans/spi/java/platform/support/ForwardingJavaPlatform.java#L36

it delegates all non-final methods to the delegate... the properties getter is final and can not be delegated.

Reply:
Damn, should have looked closer in the debugger.

My vote is to un-final so it can be properly delegated. Stupid design decision.

Reply:
maybe for NB 21? Lets get this in. I don't really like to make last minute fixes like this. The proxy was added when mx project support was introduced, it might break random things when it suddenly has properties in it.

Reply:
Yes, agreed, leave that for NB21.

Reply:
filed as #6568, if you are ok with this PR, please approve.